File: admins/git-repo/packages/yandex-corba-balancer-interface-check/balancer-interface-check.pl
#! /usr/bin/perl

use strict;
use warnings;
use Getopt::Long;
use Sys::Hostname;
use LWP::Simple;
use Socket qw( SOCK_STREAM SOCK_RAW NI_NUMERICHOST );
use Socket::GetAddrInfo qw( getaddrinfo getnameinfo );

my $DEBUG = 0;

my ($balancer, $port);
my (@ip, @err_ip);

GetOptions (
        'balancer=s' => \$balancer,
        'b=s'        => \$balancer,
        'port=i'     => \$port,
        'p=i'     => \$port,
        ) or die $!;

unless ($port and $balancer) {
        print ("Usage: $0 -b balancer_name -p port\n");
        exit;
}

my $my_hostname = hostname();
print ("my_hostname = $my_hostname\n") if $DEBUG;

sub get_balancer_ip($) {
        my $balancer = shift;
        my %hints = ( socktype => SOCK_RAW );
        my ( $err, @res ) = getaddrinfo( "$balancer", "", \%hints );
        die "Cannot resolve name - $err" if $err;

        foreach my $ai (@res) {
                my ($err, $ipaddr) = getnameinfo($ai->{addr}, NI_NUMERICHOST);
                die "Cannot getnameinfo - $err" if $err;

                push (@ip, $ipaddr);
        }

        return @ip;
}

sub get_my_vip($) {
        my $balancer = shift;
        my $vip_on_balancers = get "https://racktables.yandex.net/export/vip_on_balancers.php?vip=$balancer&vport=80";
        my @vip_on_balancers = split(/\s/, $vip_on_balancers);

        my @temp = grep (/$my_hostname/, @vip_on_balancers);
        my $my_vip;
        foreach my $str (@temp) {
                if ($str =~ m|.*\((.*)/.*\).*|g ) {
                        $my_vip = $1;
                }
        }

        return $my_vip;
}

sub check_interface($) {
        my $vip = shift;
        my $check = system ("ip a sh dev eth0 | grep $vip 2>&1 >/dev/null");

        return $check;
}

@ip = get_balancer_ip($balancer);

if ($DEBUG) {
        print ("ip: $_\n") foreach (@ip);
}
foreach my $balancer_ip (@ip) {
        my $my_vip = get_my_vip($balancer_ip);
        if (check_interface($my_vip)) {
                push (@err_ip, $my_vip);
        }
}

unless (@err_ip) {
        print ("0; Balancer ip's UP\n");
} else {
        print ("2; Balancer ip @err_ip not configured\n");
}

File: admins/git-repo/packages/yandex-corba-balancer-interface-check/balancer-interface-check.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import argparse
import socket
import urllib
import re
import subprocess

def createParser ():
    parser = argparse.ArgumentParser()
    parser.add_argument ('--balancer', '-b', required=True)
    parser.add_argument ('--port', '-p', default='80')

    return parser

parser = createParser()
namespace = parser.parse_args(sys.argv[1:])

if namespace.balancer:
	balancer = namespace.balancer
	b_port = namespace.port

hostname = socket.gethostname()
my_ipv4 = socket.gethostbyname(hostname)

def get_b_ip(balancer):
    ips = socket.getaddrinfo(balancer, 80, 0, 0, socket.SOL_TCP)
    address = []
    for i in ips:
        address.append(i[4][0])

    return address

# Get vip from racktables
def get_vip(ip, port):
	c = urllib.urlopen('https://racktables.yandex.net/export/vip_on_balancers.php?vip=' + ip + '&vport=' + port).read()
	array = c.split('\n')
	if array == ['']:
		print "2; Not found balancers in racktables"
		raise SystemExit(1)

	for line in array:
		p = re.compile(hostname + "\((.*)/.*\).*")
		aa = p.search(line)
		if aa:
			bb, cc = aa.span()
			r = re.compile (r'[\(/\)]')
			array_vip = r.split(line[1:cc])
			vip = array_vip[1]
			
			return vip

b_ips = get_b_ip(balancer)

for ip in b_ips:
	vip = get_vip(ip, b_port)
	check_int = subprocess.call("ip a sh | grep %s 2>&1 >/dev/null" % vip, shell=True)
	if check_int != 0:
		print "2; Balancer ip not configured"
		raise SystemExit(1)

print "0; Balancer ip's UP"
